Extract selectedChild tabheader functionality and enable for Manage Events

mattwire opened the pull request:
Overview
----------------------------------------
Some pages with tabs in CiviCRM allow you to link directly to a specific tab using the URL parameter `&selectedChild=tabname` (eg. for the contact summary to link directly to the contributions tab `tabname="contribute"`

This is currently supported for:
* Contact tabs
* Contribution page tabs
* Messagetemplates tabs
* Extensions tabs
* Profiles tabs

... But the code is duplicated in each template.

This PR extracts the duplicate code to a common shared template and adds support to the event page tabs.

Before
----------------------------------------
* Duplicated code in template for each instance.
* No support in event tabs.

After
----------------------------------------
* Shared code which can be included anywhere tabs are implemented.
* Support for `selectedChild` in event tabs.

Technical Details
----------------------------------------
For the events tab we have to retrieve the `selectedChild` parameter.  Other than that this is just template extraction.

Comments
----------------------------------------

mlutfy commented:
jenkins, test this please

seamuslee001 commented on CRM/Event/Form/ManageEvent/TabHeader.php:
```diff
@@ -44,6 +44,8 @@ class CRM_Event_Form_ManageEvent_TabHeader {
    * @return array
    */
   public static function build(&$form) {
+    $form->assign('selectedChild', CRM_Utils_Request::retrieve('selectedChild', 'String', $form));
```
should this be alphanumeric or is that too restrictive?

mattwire replied:
I think it probably could be, I'm not sure if there's any tabs with _ in the name

eileenmcnaughton replied:
@seamuslee001 I feel like we need another rule - Alphanumeric + a few more - hyphen, dash & possibly space

eileenmcnaughton replied:
In this case only 'user-profiles' is not Alphanumeric - & as a new assign we can be restrictive. I wonder if we can convert that to one word?

mattwire replied:
Right, Alphanumeric includes the "-" and "_" characters so we're fine to switch this to alphanumeric - I've changed all four instances and updated the PR.

eileenmcnaughton commented on templates/CRM/common/TabSelected.tpl:
```diff
@@ -0,0 +1,12 @@
+  {if $selectedChild}selectedTab = '{$selectedChild}';{/if}
+  {literal}
+  CRM.$(function($) {
+    var tabIndex = $('#tab_' + selectedTab).prevAll().length
```
I think there should be a semicolon at the end of this line -for style even if it works without

eileenmcnaughton commented:
Testing notes (still wip) 

Tested urls 

work (before & after)
civicrm/admin/uf/group?reset=1&selectedChild=reserved-profiles
civicrm/admin/uf/group?reset=1&selectedChild=user-profiles

doesn't work (before & after)
civicrm/admin/extensions?reset=1&selectedChild=addnew


@mattwire what do you think about putting this line

$this->assign('selectedChild', CRM_Utils_Request::retrieve('selectedChild', 'Alphanumeric', $this));

into CRM_Core_Basic_Page::run()

it feels like a common enough  & cheap enough thing it might make sense

eileenmcnaughton commented on templates/CRM/Admin/Page/Extensions.tpl:
```diff
@@ -101,20 +101,7 @@
-    CRM.$(function($) {
-      var tabIndex = $('#tab_' + selectedTab).prevAll().length;
-      $("#mainTabContainer").tabs({active: tabIndex});
-      $(".crm-tab-button").addClass("ui-corner-bottom");
```
class loss also affects ContributionPage

eileenmcnaughton commented on templates/CRM/Contribute/Page/Tab.tpl:
```diff
@@ -31,18 +31,8 @@
-        {* Tab management *}
-        <script type="text/javascript">
-          var selectedTab  = 'contributions';
+        {include file="CRM/common/TabSelected.tpl" defaultTab="contributions" tabContainer="#secondaryTabContainer"}
```
I couldn't figure out how to select the recurring subtab (before or after this change)

mattwire replied:
@eileenmcnaughton No, you can't :-(  But maybe once this abstraction is in place we might be able to look at it

eileenmcnaughton commented on CRM/Event/Form/ManageEvent/TabHeader.php:
```diff
@@ -235,7 +237,7 @@ public static function getCurrentTab($tabs) {
     if (is_array($tabs)) {
       foreach ($tabs as $subPage => $pageVal) {
-        if ($pageVal['current'] === TRUE) {
+        if (CRM_Utils_Array::value('current', $pageVal) === TRUE) {
```
are we sure we will not get an enotice if we remove this enotice protection?

mattwire commented:
Ok, I've addressed comments, extensions is now working.  MessageTemplates uses CRM_Core_Page instead of CRM_Core_Page_Basic so I've not abstracted out the assigns there as I nearly fell into the rabbit hole again.

eileenmcnaughton commented:
thanks @mattwire  - good to merge

seamuslee001 commented:
Test fails unrelated, merging as per the tag

jitendrapurohit commented on templates/CRM/Event/Form/ManageEvent/Tab.tpl:
```diff
@@ -107,3 +107,4 @@ CRM.$(function($) {
 </script>
 {/literal}
 {include file="CRM/Event/Form/ManageEvent/ConfirmRepeatMode.tpl" entityID=$id entityTable="civicrm_event"}
+{include file="CRM/common/TabSelected.tpl" defaultTab="settings"}
```
@mattwire @eileenmcnaughton Currently this line breaks the manage event URLs if selectedChild is not appended to it. Eg

https://dmaster.demo.civicrm.org/civicrm/event/manage/fee?reset=1&action=update&id=3 goes to the settings page instead of opening the `fee` tab. Similarly all the configuration links from the manage event page goes to the settings page.

Maybe, we should just 

- remove this line from the file or 
- append selectedChild to the links in Manage Event Page. But we already specify the tab name in the link before the query string eg `civicrm/event/manage/fee?reset=1`. 
- Format the links to look like `civicrm/event/manage?..selectedChild=fee` (might need xml/upgrade changes along with fixing the links in core).

eileenmcnaughton replied:
@mattwire @jitendrapurohit maybe we should assign defaultTab from php?

@mattwire I think this needs to be fixed in the rc

mattwire replied:
Ok. So before this PR the selectedChild stuff didn't work for events at all.  Events is the only one that specifies the tab as part of the link.

> 1. remove this line from the file

Specifying the "defaultTab" is standard across all usages of the tab selector code and is meant to be a fallback default if selectedChild is not defined.

2. append selectedChild to the links in Manage Event Page. But we already specify the tab name in the link before the query string eg civicrm/event/manage/fee?reset=1.

This is probably the easiest / safest way to resolve this, unless we assign defaultTab from PHP in 1. 

> 3. Format the links to look like civicrm/event/manage?..selectedChild=fee (might need xml/upgrade changes along with fixing the links in core).

This would be more standard and is what we should probably aspire to for consistency, but is also much more likely to cause breakage.

mattwire commented:
@jitendrapurohit @eileenmcnaughton Please see Links to manage event tabs broken following work on "selected tab" functionality.

colemanw commented:
I think the work in this PR arose from an unfortunate lack of communication. Back in 2013 I worked to get the "default tab" functionality out of the `.tpl` and into a proper javascript file: `templates/CRM/common/TabHeader.js`.  Whatever leftover code was in the template didn't need to be kept, it could just be removed in favor of the pure js approach.

See https://lab.civicrm.org/dev/core/-/issues/2215#note_50975
